controllers/machineregistration_controller.go

```diff
@@ -0,0 +1,342 @@
+	logger := ctrl.LoggerFrom(ctx)
+
+	logger.Info("Setting registration token and url")
+
+	serverURL, err := r.getRancherServerURL(ctx)
+	if err != nil {
+		return err
+	}
+
+	if mRegistration.Status.RegistrationToken == "" {
+		mRegistration.Status.RegistrationToken, err = randomtoken.Generate()
+		if err != nil {
+			return fmt.Errorf("failed to generate registration token: %w", err)
+		}
+	}
+
+	mRegistration.Status.RegistrationURL = fmt.Sprintf("%s/elemental/registration/%s", serverURL, mRegistration.Status.RegistrationToken)
+
+	return nil
+}
+
+func (r *MachineRegistrationReconciler) getRancherServerURL(ctx context.Context) (string, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	setting := &managementv3.Setting{}
+	if err := r.Get(ctx, types.NamespacedName{Name: "server-url"}, setting); err != nil {
+		return "", fmt.Errorf("failed to get server url setting: %w", err)
+	}
+
+	if setting.Value == "" {
+		err := errors.New("server-url is not set")
+		logger.Error(err, "can't get server-url")
+		return "", err
+	}
+
+	return setting.Value, nil
+}
+
+func (r *MachineRegistrationReconciler) createRBACObjects(ctx context.Context, mRegistration *elementalv1.MachineRegistration) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	logger.Info("Reconciling RBAC resources")
+
+	ownerReferences := []metav1.OwnerReference{
+		{
+			APIVersion: elementalv1.GroupVersion.String(),
+			Kind:       "MachineRegistration",
+			Name:       mRegistration.Name,
+			UID:        mRegistration.UID,
+			Controller: pointer.Bool(true),
+		},
+	}
+
+	logger.V(5).Info("Creating role")
+	if err := r.Create(ctx, &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            mRegistration.Name,
+			Namespace:       mRegistration.Namespace,
+			OwnerReferences: ownerReferences,
+		},
+		Rules: []rbacv1.PolicyRule{{
+			APIGroups: []string{""},
+			Verbs:     []string{"get", "watch", "list", "update", "patch"}, // TODO: Review permissions, does it need update, patch?
+			Resources: []string{"secrets"},
+		}, {
+			APIGroups: []string{"management.cattle.io"},
+			Verbs:     []string{"get", "watch", "list"},
+			Resources: []string{"settings"},
+		},
+		},
+	}); err != nil && !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create role: %w", err)
+	}
+
+	logger.V(5).Info("Creating service account")
+	if err := r.Create(ctx, &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            mRegistration.Name,
+			Namespace:       mRegistration.Namespace,
+			OwnerReferences: ownerReferences,
+		},
+	}); err != nil && !apierrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create service account: %w", err)
+	}
+
+	logger.V(5).Info("Creating token secret for the service account")
+	if err := r.Create(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            mRegistration.Name + "-token",
+			Namespace:       mRegistration.Namespace,
```

Comment:
It would be great to add here also elemental.cattle.io/managed: "true". It will be easier to find all objects created by elemental-operator and eventually backup them or do cleanup